examples/stress.rs

```diff
@@ -0,0 +1,828 @@
+                RoutingEvent::Demoted => self.stats.demotions += 1,
+                RoutingEvent::MessageReceived { content, dst, .. } => {
+                    let message: ProbeMessage = bincode::deserialize(&content)?;
+                    let dst = match dst {
+                        DstLocation::Section(name) => name,
+                        DstLocation::Node(name) => name,
+                        DstLocation::Direct => {
+                            return Err(format_err!("unexpected probe message dst: {:?}", dst))
+                        }
+                    };
+
+                    self.probe_tracker.receive(&dst, message.proof_share);
+                }
+                _ => {
+                    // Currently ignore the other event variants. This might change in the future,
+                    // if we come up with something interesting to use those events for.
+                }
+            },
+        }
+
+        self.try_print_status();
+
+        Ok(())
+    }
+
+    fn new_node_id(&mut self) -> u64 {
+        let id = self.next_id;
+        self.next_id = self.next_id.checked_add(1).expect("too many nodes");
+        id
+    }
+
+    // Returns the socket addresses to bootstrap against.
+    async fn get_bootstrap_addrs(&self) -> Result<Vec<SocketAddr>> {
+        // Number of bootstrap contacts to use. Use more than one to increase the chance of
+        // successful bootstrap in case some of the bootstrap nodes get dropped.
+        //
+        // FIXME: seems there is a bug which causes a hang if more than one contact is used.
+        // Need to investigate.
+        const COUNT: usize = 1;
+
+        // Use the oldest nodes in the network as the bootstrap contacts.
+        future::try_join_all(
+            self.nodes
+                .values()
+                .filter_map(|node| match node {
+                    Node::Joined { node, age, .. } => Some((node, age)),
+                    Node::Joining => None,
+                })
+                .sorted_by(|(_, lhs_age), (_, rhs_age)| lhs_age.cmp(rhs_age).reverse())
+                .take(COUNT)
+                .map(|(node, _)| node.our_connection_info()),
+        )
+        .await
+        .map_err(Into::into)
+    }
+
+    // Send messages to probe network health.
+    async fn send_probes(&mut self) -> Result<()> {
+        // Cache the (src, dst) pairs of already sent messages.
+        let mut cache = BTreeMap::new();
+
+        let nodes = self.nodes.values().filter_map(|node| match node {
+            Node::Joined {
+                node, name, prefix, ..
+            } => Some((node, name, prefix)),
+            Node::Joining => None,
+        });
+
+        for (node, name, prefix) in nodes {
+            let dst = *cache
+                .entry(prefix)
+                .or_insert_with(|| prefix.substituted_in(rand::random()));
+            self.try_send_probe(node, *name, dst).await?;
+            self.probe_tracker.send(*prefix, dst);
```

Comment:
Adult nodes will also trigger the probe_tracker to send once, is it ok ?

Reply:
Hm, probably not ideal. Say a split happens and some non-elder nodes are lagging behind and still think they are in the pre-split section. Then those will cause the counter of the sent probes to increase for the pre-split section even though no messages are actually being sent there. Good catch! Will fix.